File: PyLabFastAPI/app/schemas/course.py
# PyLabFastAPI/app/schemas/course.py
from pydantic import BaseModel
from typing import Optional, List
from datetime import datetime


# === 课程相关 ===
class CourseBase(BaseModel):
    title: str
    desc: Optional[str] = None
    cover: Optional[str] = None  # 封面图URL
    price: float = 0


class CourseCreateReq(CourseBase):
    pass


class CourseOut(CourseBase):
    id: int
    teacher_id: int
    is_published: bool
    created_at: datetime
    view_count: int = 0

    # [新增] 当前用户是否已加入 (仅在登录且请求详情时有效)
    is_joined: bool = False

    class Config:
        from_attributes = True


# === 章节相关 ===
class ChapterCreateReq(BaseModel):
    title: str
    rank: int = 0  # 排序

from enum import Enum

# === 课时相关 ===
class LessonType(str, Enum):
    VIDEO = "video"
    PROBLEM = "problem"

class LessonCreateReq(BaseModel):
    title: str
    type: LessonType  # 必须是 'video' 或 'problem'
    rank: int = 0
    # 如果是视频课，必须传 video_id
    video_id: Optional[int] = None
    # 如果是题目课，必须传 problem_id (后面开发OJ时用)
    problem_id: Optional[int] = None


class LessonOut(BaseModel):
    id: int
    title: str
    type: str
    rank: int
    video_id: Optional[int] = None

    # [新增] 必须把题目ID返回给前端
    problem_id: Optional[int] = None

    class Config:
        from_attributes = True


class ChapterOut(BaseModel):
    id: int
    title: str
    rank: int

    # [新增] 嵌套返回该章节下的所有课时
    lessons: List[LessonOut] = []

    class Config:
        from_attributes = True


class UserCourseOut(BaseModel):
    id: int
    course_id: int
    user_id: int
    progress: float
    joined_at: datetime

    class Config:
        from_attributes = True

class CourseUpdateReq(BaseModel):
    title: str = None
    desc: str = None
    cover: str = None
    price: float = None
    is_published: bool = None

class ChapterUpdateReq(BaseModel):
    title: str
    rank: int

class LessonUpdateReq(BaseModel):
    title: str
    type: str
    rank: int
    video_id: int = None

File: PyLabVUE/src/api/course.ts
// src/api/course.ts
import request from '@/utils/request'
import type { ApiResponse } from '@/model/common'

// === 接口定义 ===
export interface CourseCreateReq {
  title: string
  desc?: string
  cover?: string
  price?: number
}

export interface ChapterCreateReq {
  title: string
  rank?: number
}

export interface LessonCreateReq {
  title: string
  type: 'video' | 'problem'
  rank?: number
  video_id?: number
}

export interface Course {
  id: number
  title: string
  desc?: string
  cover: string
  price: number
  view_count: number
  is_published: boolean // [新增]
  teacher_name?: string
  created_at: string
  is_joined?: boolean
}

export interface Lesson {
  id: number
  title: string
  type: 'video' | 'problem'
  rank: number
  video?: {
    play_url: string
    duration: number
    title: string
    file_key: string
    file_size: number
  }
  problem?: {
    id: number
    title: string
    content: string
    init_code: string
    time_limit: number
    memory_limit: number
  }
  problem_id?: number
}

export interface Chapter {
  id: number
  title: string
  rank: number
  lessons: Lesson[]
}

export interface UserCourse {
  id: number
  course_id: number
  user_id: number
  progress: number
  joined_at: string
}

export interface CourseListQuery {
  page?: number
  size?: number
  keyword?: string
  sort?: 'new' | 'hot'
}

export interface CourseListRes {
  items: Array<Course & { teacher_name: string }>
  total: number
  page: number
  size: number
}

export interface CourseDetailRes {
  info: Course
  related: Course[]
}

export class CourseApi {
  static async createCourse(data: CourseCreateReq) {
    const res = await request.post<ApiResponse<Course>>('/courses', data)
    return res.data.data
  }

  // [修复] 补全 getMyCourses
  static async getMyCourses() {
    const res = await request.get<ApiResponse<Course[]>>('/courses/my')
    return res.data.data
  }

  static async createChapter(courseId: number, data: ChapterCreateReq) {
    const res = await request.post<ApiResponse<Chapter>>(`/courses/${courseId}/chapters`, data)
    return res.data.data
  }

  static async createLesson(courseId: number, chapterId: number, data: LessonCreateReq) {
    const res = await request.post<ApiResponse<Lesson>>(
      `/courses/${courseId}/chapters/${chapterId}/lessons`,
      data,
    )
    return res.data.data
  }

  static async getCourses(params: CourseListQuery) {
    const res = await request.get<ApiResponse<CourseListRes>>('/courses', { params })
    return res.data.data
  }

  static async getCourseDetail(courseId: number) {
    const res = await request.get<ApiResponse<CourseDetailRes>>(`/courses/${courseId}`)
    return res.data.data
  }

  static async joinCourse(courseId: number) {
    const res = await request.post<ApiResponse<UserCourse>>(`/courses/${courseId}/join`)
    return res.data.data
  }

  static async getCourseChapters(courseId: number) {
    const res = await request.get<ApiResponse<Chapter[]>>(`/courses/${courseId}/chapters`)
    return res.data.data
  }

  static async publishCourse(courseId: number) {
    const res = await request.patch<ApiResponse<any>>(`/courses/${courseId}`, {
      is_published: true,
    })
    return res.data.data
  }

  static async updateCourse(id: number, data: Partial<Course>) {
    return request.patch(`/courses/${id}`, data)
  }

  static async updateChapter(id: number, data: { title: string; rank: number }) {
    return request.put(`/courses/chapters/${id}`, data)
  }

  static async deleteChapter(id: number) {
    return request.delete(`/courses/chapters/${id}`)
  }

  static async updateLesson(id: number, data: any) {
    return request.put(`/courses/lessons/${id}`, data)
  }

  static async deleteLesson(id: number) {
    return request.delete(`/courses/lessons/${id}`)
  }
  static async deleteCourse(id: number) {
    return request.delete(`/courses/${id}`).then((res) => res.data)
  }
}
